### PR TITLE
[web-animations-1] Eliminate usage of CSSPseudoElement

### PR DESCRIPTION
**Pull request by george-steel**
Fix issue #4301

### CODE REVIEW

**Comment by george-steel** [comment]
@stephenmcgruer @birtles here's my first draft. I'm still not sure what to do about [commitStyles](https://drafts.csswg.org/web-animations-1/#dom-animation-commitstyles) as the DOM has no notion of inline style for pseudo-elements.

**Comment by birtles** [comment]
> I'm still not sure what to do about [commitStyles](https://drafts.csswg.org/web-animations-1/#dom-animation-commitstyles) as the DOM has no notion of inline style for pseudo-elements.

I'm not sure you need to do anything? `commitStyles` [throws](https://drafts.csswg.org/web-animations-1/#commit-computed-styles) if any of the targets is a pseudo-element.

File: web-animations-1/Overview.bs [hunk 1]
```diff
@@ -3188,9 +3189,14 @@ Keyframe effects {#keyframe-effects}
...
-an element or pseudo-element such as <code>::before</code> or
-<code>::after</code> [[!SELECT]] referred to as the <dfn>target element</dfn>.
+effect</a> that use the output of the timing model to update CSS properties of
+an element known as the <dfn>target element</dfn> or one of its <a>pseudo-elements</a>
+(in which case the target element is the <a>originating element</a> of the pseudo-element).
+
+The combination of <a>target element</a> and <a>pseudo-element</a> selector
+referred to as the <dfn>effect target</dfn>. In the case that the effect target does not exist
+(for example when it targets a `::after` pseudo-element where one does not exist)
+the effect applies no styles until the target comes into existence.
```
Comment (birtles) [inline web-animations-1/Overview.bs]: I'd rather we don't change the meaning of "target element". This should be an API-only change, the model itself should stay the same, right?

Comment (birtles) [reply web-animations-1/Overview.bs]: Actually, may it's ok. I'd probably just prefer to reword this so that the definition of "effect target" comes before "target element".

So the first paragraph should be just:

> Keyframe effects are a kind of animation effect that use the output of the timing model to update CSS properties for
an element or pseudo-element such as ::before or ::after [[!SELECT]] referred to as the effect target.

File: web-animations-1/Overview.bs [hunk 2]
```diff
@@ -3290,7 +3296,7 @@ To produce <a>computed keyframe offsets</a>, we define a procedure to
 
 <a>Computed keyframes</a> are produced using the following procedure.
 Note that this procedure is only performed on a <a>keyframe effect</a> having
-a <a>target element</a> for which computed property values can be calculated.
+a existing <a>effect target</a> for which computed property values can be calculated.
```
Comment (birtles) [inline web-animations-1/Overview.bs]: I think if you're going to introduce "existing" here, I think you need to link to a definition of what it means for a target to exist.

File: web-animations-1/Overview.bs [hunk 3]
```diff
@@ -3236,7 +3242,7 @@ Before calculating the <a>effect value</a> of a <a>keyframe effect</a>,
 the property values specified on its <a>keyframes</a> are resolved to
 <a>computed values</a>, and the offset to use for any keyframes with a null
 <a>keyframe offset</a> is computed. The result of resolving these values is
-a set of <dfn>computed keyframes</dfn>.
+a set of <dfn>computed keyframes</dfn> for a particular moment in time.
```
Comment (birtles) [inline web-animations-1/Overview.bs]: Not sure this addition is needed?

File: web-animations-1/Overview.bs [hunk 4]
```diff
@@ -3357,7 +3363,7 @@ given <var>iteration progress</var>, <var ignore=''>current iteration</var> and
 1.  If <a>animation type</a> of the <var>target property</var> is
     <a>not animatable</a> abort this procedure
     since the effect cannot be applied.
-1.  Define the <dfn>neutral value for composition</dfn> as a value
+1.  Define the <dfn>identity value for composition</dfn> as a value
```
Comment (birtles) [inline web-animations-1/Overview.bs]: Is this change related?

File: web-animations-1/Overview.bs [hunk 5]
```diff
@@ -3301,24 +3307,24 @@ a <a>target element</a> for which computed property values can be calculated.
...
         For logical properties [[!CSS-LOGICAL-1]], add the [=equivalent physical
         properties=] [[!CSS-WRITING-MODES-4]] based on the computed value of
-        'writing-mode' and/or 'direction' for <a>target element</a>.
+        'writing-mode' and/or 'direction' for the <a>effect target</a>.
 
-        For example, if <var>keyframe</var> has a value of &lsquo;12pt&rsquo;
+        For example, if <var>keyframe</var> has a value of ‘12pt’
         for the 'border-width' property, the user agent may produce a computed
-        value of &lsquo;16px&rsquo; for each of the longhand properties:
+        value of ‘16px’ for each of the longhand properties:
```
Comment (birtles) [inline web-animations-1/Overview.bs]: Should these changes be a separate patch?

File: web-animations-1/Overview.bs [hunk 6]
```diff
@@ -4788,11 +4796,17 @@ interface KeyframeEffect : AnimationEffect {
...
+::  The <a>target element</a> being animated by this object
+    (either the <a>effect target</a> if it is an element or its
+    <a>originating element</a> if it is a pseudo-element).
+    This may be `null` for animations that do not target
     a specific element such as an animation that produces a sound
     using an audio API.
 
+:   <dfn attribute for=KeyframeEffect>pseudoElement</dfn>
+::  The <a>effect target</a>'s <a>pseudo-element</a> selector if the effect target is a pseudo-element.
+    This is `null` if the <a>effect target</a> is an element or has no specific target.
```
Comment (birtles) [inline web-animations-1/Overview.bs]: Does this need to define that this is `::before` or `::after` or `::marker`? i.e. both the specific syntax used (not the old-style `:before` etc.) and the set of pseudo elements supported?

File: web-animations-1/Overview.bs [hunk 7]
```diff
@@ -5743,9 +5762,9 @@ animation.play();</pre>
         <em>and</em>
     * Its [=replace state=] is <em>not</em> [=removed replace state|removed=].
 
-    For the purposes of this method
-    a [=pseudo-element=] is considered to be
-    a [=child=] of its [=originating element=].
+    Note: Since an {{Animation}} targeting a <a>pseudo-element</a> has the
+    <a>originating element</a> as its <a>target element</a>, calling {{Animatable/getAnimations}}
+    on an element will return <a>animations</a> targeting both the element itself and its pseudo-elements.
```
Comment (birtles) [inline web-animations-1/Overview.bs]: It should only return them if `subtree: true` is set. I think the want to keep roughly the original spec text here.

**Comment by birtles** [comment]
I've done a first pass review for now. It's looking pretty good but I'd like to have another look after those changes have been made. It would also be easier to review if we can move the cosmetic and unrelated changes to a separate PR / patch.

File: web-animations-1/Overview.bs (hunk 5 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: Perhaps you're right and this and the previous change should go in a separate cleanup patch.

File: web-animations-1/Overview.bs (hunk 3 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: I was trying to clarify something that got misread and lead to bugs in our current implementation.

File: web-animations-1/Overview.bs (hunk 2 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: Existing means present in the layout tree, such as a `::before` with a content property, as opposed to one without content that doesn't appear. The idea is to hang Animations off of their target Elements which have a well-defined lifecycle and make Animation lifecycle independent of the ill-defined pseudo-element lifecycle. The Animation always runs and has a visible effect iff the pseudo-element does.

File: web-animations-1/Overview.bs (hunk 6 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: My idea was that this since there's now the notion of a target Element to hang the Animation off, we could apply animations to all valid pseudo-elements: both the tree-abiding ones, ones that act as alternate style slots (such as ::first-line, ::selection, only supporting the properties those selectors support), and shadow part styles. I see the restriction to before after and marker as an accidental effect of needing a layout object as the effect target.

File: web-animations-1/Overview.bs (hunk 7 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: See my above comments about hanging Animations off of elements.

Comment (george-steel) [reply web-animations-1/Overview.bs]: One alternative proposed by @stephenmcgruer if to add a `pseudos` option to select whether just to get Animations targeting the element's base style or whether to get Animations targeting target pseudo-elements as well (i.e. all Animations hanging off the element).

File: web-animations-1/Overview.bs (hunk 1 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: Good idea to reorder.

File: web-animations-1/Overview.bs (hunk 6 above)
Comment (birtles) [reply web-animations-1/Overview.bs]: That sounds like a separate issue to this PR which requires further discussion. (I know Gecko has no intention of implementing animations on a wider range of pseudo-elements at this stage.)

For now we still need to define / link to the appropriate syntax here.

File: web-animations-1/Overview.bs (hunk 7 above)
Comment (birtles) [reply web-animations-1/Overview.bs]: Let's keep the existing behavior for now. If you'd like to change it, let's use a separate issue / PR for that.

File: web-animations-1/Overview.bs (hunk 3 above)
Comment (birtles) [reply web-animations-1/Overview.bs]: Can you elaborate on that? I suspect there's probably a better way to clarify in that case.

File: web-animations-1/Overview.bs (hunk 2 above)
Comment (birtles) [reply web-animations-1/Overview.bs]: That still sounds like existing wants some sort of definition.

File: web-animations-1/Overview.bs (hunk 6 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: Are you saying we should add a restriction of which pseudo-elements to support? Currently there is no such restriction in the spec other than the fact that CSSPseudoElement:EventTarget is too much of a mess to even come close to implementing in all but those cases.

Specifying that is returns pseudo-elements in the modern syntax is a good idea.

File: web-animations-1/Overview.bs (hunk 3 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: I think there is, I plan on proposing an (isomorphic) refactor to this section to clarify it as a separate patch. Perhaps this should wait until then. The idea was that this depends on what the inherited styles are at when they're calculated, which must be the same time they're used (i.e. a keyframes do not capture parent styles at the moment of their creation).

File: web-animations-1/Overview.bs (hunk 7 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: Ok, it's a bit odd, but having a separate issue for that makes sense. I was trying to keep the invariant of it returns everything hanging off that element instead of the invariant of iit only returning the same set of animations as before.

File: web-animations-1/Overview.bs (hunk 2 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: Maybe just remove "existing" since "for which computed property values can be calculated" does the job. As this procedure is affine, it can always be skipped when there is nothing to style.

File: web-animations-1/Overview.bs (hunk 6 above)
Comment (birtles) [reply web-animations-1/Overview.bs]: No, no need to an explicit restriction.

File: web-animations-1/Overview.bs (hunk 3 above)
Comment (birtles) [reply web-animations-1/Overview.bs]: Ok that makes sense. And it probably would be best as a separate refactor.

File: web-animations-1/Overview.bs (hunk 2 above)
Comment (birtles) [reply web-animations-1/Overview.bs]: Ok, makes sense.

**Comment by george-steel** [comment]
@birtles I made the changes and split out the cosmetic stuff (it's no longer in this PR).

File: web-animations-1/Overview.bs [hunk 8]
```diff
@@ -3799,14 +3805,14 @@ if <em>all</em> of the following conditions are true:
     [=timeline=].
 *   The [=animation=] has an associated [=target effect=].
 *   The [=target effect=] associated with the [=animation=] is [=in effect=].
-*   The [=target effect=] has an associated [=target element=].
+*   The [=target effect=] has an associated [=effect target=].
```
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply web-animations-1/Overview.bs]: @birtles for your thoughts here

[the remark this answers is not in the document]
Comment (george-steel) [reply web-animations-1/Overview.bs]: I agree this is akward, your solution looks good to me.

**Comment by george-steel** [comment]
Sorry about the force push, A bunch of brian's changes required splitting commits between this PR and a future one.

File: web-animations-1/Overview.bs [hunk 9]
```diff
@@ -3187,10 +3188,15 @@ the <a>animation type</a> is <a>discrete</a>.
...
-<code>::after</code> [[!SELECT]] referred to as the <dfn>target element</dfn>.
+<dfn lt="keyframe effect">Keyframe effects</dfn> are a kind of [=animation
+effect=] that use the output of the timing model to update CSS properties of
+an element or [=pseudo-element=] (such as `::before` or `::after` [[!select]])
+referred to as the <dfn>effect target</dfn>.
+
+In order for the [=effect target=] to have a well-defined lifecycle,
+it is reperesented as an {{Element}} known as the <dfn>target element</dfn>
+(if the target is a [=pseudo-element=] this is the target's [=originating element=])
+and a [=pseudo-element=] selector (which is `null` if the [=effect target=] is an element).
```
[the remark this answers is not in the document]
Comment (george-steel) [reply web-animations-1/Overview.bs]: done

File: web-animations-1/Overview.bs (hunk 8 above)
[the remark this answers is not in the document]
Comment (birtles) [reply web-animations-1/Overview.bs]: I think `effect` is too vague a term for the model (it's fine for the API though). Any old key frame effect is an effect so this can easily get confusing. We mean to say the particular effect that the animation is playing.

I agree we should keep the term `effect target` but rename `target effect` to something else.

At one point I think we called the `target effect` the `source effect` but that's slightly inaccurate. An animation _drives_ an effect, but is quite happy to keep ticking along without one. So it's not really the "source" for anything.

`client effect`? `applied effect`? `associated effect`?

I think I like "associated effect" best?

File: web-animations-1/Overview.bs [hunk 10]
```diff
@@ -4788,11 +4795,21 @@ interface KeyframeEffect : AnimationEffect {
...
     a specific element such as an animation that produces a sound
     using an audio API.
 
+:   <dfn attribute for=KeyframeEffect>pseudoElement</dfn>
+::  The [=effect target=]'s [=pseudo-element=] selector.
+    This is `null` if the [=effect target=] is an element the effect
+    has or has no specific target.
+    When not `null`, this must be a valid [=pseudo-element=] selector
+    and will always return the modern syntax
+    (i.e. it will always reutrn `::before` and never `:before`).
```
[the remark this answers is not in the document]
Comment (birtles) [reply web-animations-1/Overview.bs]: Yes, we need something better than "modern syntax". Ideally a link to wherever the syntax is defined and a example in braces (e.g. `::before`) so implementers and authors alike know what to expect. I seem to remember dbaron encouraging me to spec that at some point which might be why CSSPseudoElement has the syntax explicitly listed:

  https://drafts.csswg.org/css-pseudo-4/#CSSPseudoElement-interface

**Comment by birtles** [comment]
> We seem to have lost Brian's earlier comments (possibly due to a force-push?), so please make sure to check the history and see that we don't lose track of anything that wasn't yet resolved.

Sorry, I should have done a proper review. I kept thinking, "just one more comment!"

File: web-animations-1/Overview.bs [hunk 11]
```diff
@@ -3187,10 +3188,19 @@ the <a>animation type</a> is <a>discrete</a>.
...
-effect</a> that use the output of the timing model to update CSS properties for
-an element or pseudo-element such as <code>::before</code> or
-<code>::after</code> [[!SELECT]] referred to as the <dfn>target element</dfn>.
+<dfn lt="keyframe effect">Keyframe effects</dfn> are a kind of [=animation
+effect=] that use the output of the timing model to update CSS properties of
+an element or [=pseudo-element=] (such as `::before` or `::after` [[!select]])
+referred to as the <dfn>effect target</dfn>.
+
+In order for the [=effect target=] to have a well-defined lifecycle,
+it is reperesented as an {{Element}} known as the <dfn>target element</dfn>
```
Comment (stephenmcgruer) [inline web-animations-1/Overview.bs]: s/reperesented/represented

File: web-animations-1/Overview.bs [hunk 12]
```diff
@@ -3187,10 +3188,19 @@ the <a>animation type</a> is <a>discrete</a>.
...
+and a [=pseudo-element=] selector knows as the <dfn>target pseudo-selector</dfn>.
+If the [=effect target=] is an {{Element}}, the [=target element=] is that
+element and the [=target pseudo-selector=] is `null`.
+If the [=effect target=] is a [=pseudo-element=], the [=target element=] is
+its [=originating element=], and the [=target pseudo-selector=] is as required
+to specify thatparticular [=pseudo-element=].
```
Comment (stephenmcgruer) [inline web-animations-1/Overview.bs]: s/thatparticular/that particular

File: web-animations-1/Overview.bs [hunk 13]
```diff
@@ -3187,10 +3188,19 @@ the <a>animation type</a> is <a>discrete</a>.
...
+an element or [=pseudo-element=] (such as `::before` or `::after` [[!select]])
+referred to as the <dfn>effect target</dfn>.
+
+In order for the [=effect target=] to have a well-defined lifecycle,
+it is reperesented as an {{Element}} known as the <dfn>target element</dfn>
+and a [=pseudo-element=] selector knows as the <dfn>target pseudo-selector</dfn>.
+If the [=effect target=] is an {{Element}}, the [=target element=] is that
+element and the [=target pseudo-selector=] is `null`.
+If the [=effect target=] is a [=pseudo-element=], the [=target element=] is
+its [=originating element=], and the [=target pseudo-selector=] is as required
```
Comment (stephenmcgruer) [inline web-animations-1/Overview.bs]: Super nit; the comma on this line is not needed.

File: web-animations-1/Overview.bs (hunk 8 above)
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply web-animations-1/Overview.bs]: `associated effect` SGTM. It's somewhat vague, but good enough I think.

File: web-animations-1/Overview.bs (hunk 10 above)
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply web-animations-1/Overview.bs]: https://www.w3.org/TR/selectors-4/#pseudo-element-syntax is the syntax, right?

File: web-animations-1/Overview.bs [hunk 14]
```diff
@@ -3807,9 +3807,9 @@ if <em>all</em> of the following conditions are true:
     [=removed replace state|removed=].
 *   The [=animation=] is associated with a [=monotonically increasing=]
     [=timeline=].
-*   The [=animation=] has an associated [=target effect=].
-*   The [=target effect=] associated with the [=animation=] is [=in effect=].
-*   The [=target effect=] has an associated [=effect target=].
+*   The [=animation=] has an [=associated effect=].
+*   The [=animation=]'s [=associated effect=] is [=in effect=].
+*   The [=associated effect=] has an [=effect target=].
```
Comment (birtles) [inline web-animations-1/Overview.bs]: (Nit: I guess for consistency we should make the last bullet point also say `The [=animation=]'s [=associated effect=]...`.)

File: web-animations-1/Overview.bs [hunk 15]
```diff
@@ -4809,10 +4809,10 @@ interface KeyframeEffect : AnimationEffect {
 :   <dfn attribute for=KeyframeEffect>pseudoElement</dfn>
 ::  The [=target pseudo-selector=].
     This is `null` if the [=effect target=] is an {{Element}} or is itself `null`.
-    When not `null`, this must be a valid [=pseudo-element=] selector,
-    attempting to set it to an invalid value will have no effect.
-    and will always return using the the double-colon syntax
-    (i.e. always `::before` and never `:before`).
+    When not `null`, this must be a [=pseudo-element=] selector with valid 
+    [[selectors-4#pseudo-element-syntax]].
```
Comment (birtles) [inline web-animations-1/Overview.bs]: Just curious, how does this actually render? Does it say, "...a pseudo element selector with valid syntax"?

File: web-animations-1/Overview.bs [hunk 16]
```diff
@@ -4809,10 +4809,10 @@ interface KeyframeEffect : AnimationEffect {
...
+    This will not accept invalid values and will always return the selector in
+    its canonical form (i.e. `::before` instead of `:before`).
```
Comment (birtles) [inline web-animations-1/Overview.bs]: I think this isn't strict enough. The linked-to syntax section still allows the single colon format whereas I think we want to explicitly reject that for this API since we don't want to propagate that form any further.

File: web-animations-1/Overview.bs [hunk 17]
```diff
@@ -772,10 +772,10 @@ analogous to that of a DVD player and a DVD.
 </div>
 
 An <dfn id="concept-animation">animation</dfn> connects a single <a>animation
-effect</a>, called its <dfn>target effect</dfn>, to a <a>timeline</a> and
+effect</a>, called its <dfn>associated effect</dfn>, to a <a>timeline</a> and
 provides playback control.
 Both of these associations are optional and configurable such that
-an <a>animation</a> may have no associated <a>target effect</a> or
+an <a>animation</a> may have no associated <a>associated effect</a> or
```
[the remark this answers is not in the document]
Comment (george-steel) [reply web-animations-1/Overview.bs]: I thought I got all of those. Must have missed that one.

File: web-animations-1/Overview.bs (hunk 16 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: I was aiming for consistency with other specs which accept pseudo-element selectors. Most I've encountered so far will accept but never return the legacy syntax. Should the question of whether to change that convention be handled here or should that be a separate issue/PR?

File: web-animations-1/Overview.bs (hunk 15 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: oops, I accidentally had it include the section number, fixing that now.

File: web-animations-1/Overview.bs [hunk 18]
```diff
@@ -847,14 +847,14 @@ becomes <a lt="idle play state">idle</a>, is when the procedure to
 
 </div>
 
-<h4 id="setting-the-target-effect">Setting the target effect of an
+<h4 id="setting-the-associated-effect">Setting the associated effect of an
```
[the remark this answers is not in the document]
Comment (george-steel) [reply web-animations-1/Overview.bs]: Unfortunately there isn't, I'm changing the id back for now.

File: web-animations-1/Overview.bs [hunk 19]
```diff
@@ -1413,26 +1413,26 @@ Graphically, the effect of limiting the current time is shown below.
        alt="The effect of limiting the current time of an animation.">
   <figcaption>
     The effect of limiting the <a>current time</a> of an <a>animation</a>
-    with a start time of 1s, a target effect of length 3s, and
+    with a start time of 1s, a associated effect of length 3s, and
```
[the remark this answers is not in the document]
Comment (george-steel) [reply web-animations-1/Overview.bs]: The hazards of running a regex. Thanks for catching this.

File: web-animations-1/Overview.bs (hunk 15 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: Now it renders like `a pseudo element selector with [=valid syntax=]`

File: web-animations-1/Overview.bs (hunk 16 above)
Comment (george-steel) [reply web-animations-1/Overview.bs]: @stephenmcgruer

File: web-animations-1/Overview.bs (hunk 18 above)
[the remark this answers is not in the document]
Comment (george-steel) [reply web-animations-1/Overview.bs]: @stephenmcgruer What do you think?

File: web-animations-1/Overview.bs (hunk 16 above)
Comment (birtles) [reply web-animations-1/Overview.bs]: Yes, please file a separate issue for this. I seem to remember getting advice at one point that we shouldn't accept the old syntax either.

File: web-animations-1/Overview.bs (hunk 18 above)
[the remark this answers is not in the document]
Comment (birtles) [reply web-animations-1/Overview.bs]: The [bikeshed docs](https://tabatkins.github.io/bikeshed/#id-gen) say:
> If a heading changed significantly, so that you want to change the ID, but you want links to the old heading ID to still work, put the old ID in an oldids="" attribute on the heading element. If there are multiple, comma-separate them.

**Comment by george-steel** [comment]
fixed the oldids

**Comment by stephenmcgruer** [comment]
Awesome! Can you file the issue for disallowing the old syntax @george-steel , if you haven't already?

Looking forward to the WPT test changes to accompany this too :)

**Comment by george-steel** [comment]
@birtles I'm not quite sure about your reasoning for disallowing the old syntax in this one specific case. Since it was your idea, would you mind filing the issue?

**Comment by birtles** [comment]
> @birtles I'm not quite sure about your reasoning for disallowing the old syntax in this one specific case. Since it was your idea, would you mind filing the issue?

Filed #4472.

File: web-animations-1/Overview.bs [hunk 20]
```diff
@@ -3187,10 +3187,19 @@ the <a>animation type</a> is <a>discrete</a>.
...
-an element or pseudo-element such as <code>::before</code> or
-<code>::after</code> [[!SELECT]] referred to as the <dfn>target element</dfn>.
+<dfn lt="keyframe effect">Keyframe effects</dfn> are a kind of [=animation
+effect=] that use the output of the timing model to update CSS properties of
+an element or [=pseudo-element=] (such as `::before` or `::after` [[!select]])
+referred to as the <dfn>effect target</dfn>.
+
+In order for the [=effect target=] to have a well-defined lifecycle,
+it is represented as an {{Element}} known as the <dfn>target element</dfn>
+and a [=pseudo-element=] selector knows as the <dfn>target pseudo-selector</dfn>.
```
Comment (Loirooriol) [inline web-animations-1/Overview.bs]: Typo: knows -> known
